build: containerize deployment

epuertat opened the pull request:
## Relevant Changes
* Build and deploy with `docker-compose`.
* Use `make` to wrap and launch `docker-compose` commands.
  * Targets follow the `docker-compose` commands: `make build`, `make up`, `make logs`, `make down`.
  * It dynamically sets multiple metadata like container `LABELS` (mimicking [`io.k8s.*`](https://github.com/projectatomic/ContainerApplicationGenericLabels/blob/master/vendor/redhat/labels.md)):
  	* `io.ceph.compoment`
  	* `io.ceph.summary`
  	* `io.ceph.description`
  	* `io.ceph.url`
  	* `io.ceph.version`
  	* `io.ceph.maintainer`
  	* `io.ceph.git.repo`
  	* `io.ceph.git.branch`
  	* `io.ceph.git.commit` 
* Use [`pdm`](https://pdm.fming.dev/) as Python package manager to install packages, as it supports (the recently rejected) [PEP-582](https://peps.python.org/pep-0582/) (installing dependencies to `__pypackages__` instead of using `virtualenv`). Despite the PEP rejection, it still makes sense, since containerization renders virtualenv unnecessary. 
* Create 4 containers:
  * [`ceph/spdk`](https://quay.io/repository/ceph/spdk) (`Dockerfile.spdk`): an intermediate container resulting from building spdk RPM packages and installing those on top of a [`ubi9`](https://catalog.redhat.com/software/containers/ubi9/ubi/615bcf606feffc5384e8452e) base layer. 
  * [`ceph/nvmeof`](https://quay.io/repository/ceph/nvmeof) (`Dockerfile`): the Python code consuming the SPDK JSON RPC API.
  * [`ceph/nvmeof-cli`](https://quay.io/repository/ceph/nvmeof-cli) (`Dockerfile`): a containerized CLI tool to manage the `nvmeof` gateway.
  * [`ceph/vstart-cluster`](https://quay.io/repository/ceph/vstart-cluster) (`Dockerfile.ceph`): a vstart Ceph cluster based on the [`ceph/daemon`](https://quay.io/repository/ceph/daemon) image.
* From #134 : `nvmeof-cli` reads input options from command-line arguments instead of `ceph-nvmeof.conf` file.
* Removed dependencies from `/usr/libexec` since according to [FHS](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s07.html) this is meant for internal usage.
* Add .dockerignore
* Update spdk submodule settings, ignored when dirty
* Move `proto` file to `control/proto` as it makes it easier to generate Python code

## Tasks
- [x] Containerize and isolate spdk build (~ubi8 and~ ubi9)
- [x] Containerize nvme-of build (ubi9 ~although 8 would be trivial~)
- [x] Manage Python package dependencies and metadata (pdm)
- [x] Remove dependencies from /usr/libexec (since those are not intended for external usage).
- [x] Simplify build & deployment (docker-compose instead of make)
- [x] Containerize Ceph cluster 
- [x] [Moved to #141]
  - [x] ~~Use non-privileged user inside (remove Huge Pages, via DPDK `--no-huge`). Pending of https://github.com/spdk/spdk/issues/2973 and https://github.com/ceph/ceph-nvmeof/pull/103~~
  - [x] ~~Submit fixes to spdk's `/lib/env_dpdk/init.c` for params: `--base-virtual`, `--match-allocations` and `--file-prefix`: https://review.spdk.io/gerrit/c/spdk/spdk/+/17734~~
- [x] Replace existing Makefile commands with `pdm` and `docker-compose`-based commands
- [x] Update documentation for developers and users
- [x] Update `authors` metadata in pyproject to include relevant contributors.
- [x] Pass `ceph-nvmeof.conf` as an external file rather than copied inside the container.
- [ ] Devel containers:
  - [x] Enable `nvmeof-devel` mounting the current dir inside the container `$APPDIR`.
  - [ ] Same for spdk, to allow binding it to a local build of librados and or librdb (.so? RPM?).
- [x] Replace Ceph container approach (ceph-daemon as base layer) with one installing RPMs from Shaman. This allows to specify a given Ceph version (`17.2.6`) or even a git hash, while the ceph-daemon container only allows for major release codenames (`reef`, `quincy`, etc.)
- [x] Update README with latest changes.
- [x] Bring Github CI from @baum's PR #134.

## Instructions

See [`README.md`](https://github.com/epuertat/ceph-nvmeof/blob/build-container/README.md) in this PR.

epuertat committed:
build: containerize deployment

- [x] Containerize and isolate spdk build (ubi8 and ubi9)
- [x] Containerize nvme-of build (ubi9)
- [x] Manage Python package dependencies and metadata (pdm)
- [x] Remove dependencies from /usr/libexec (since those are not intended
  for external usage).
- [x] Simplify build & deployment (docker-compose instead of make)
- [x] Add containerized Ceph cluster
- [x] Update Makefile
- [x] Update docs

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

baum commented on docker-compose.yaml:
```diff
@@ -0,0 +1,70 @@
+version: "3.8"
+services:
+  ceph-spdk:
```
```suggestion
  # Fix ownership of Ceph conf directory
  # ceph-cluster         | ./vstart.sh: line 670: /etc/ceph/ceph.conf: Permission denied
  # By default, the volume directory is owned by user 'root'
  # ceph-cluster         | drwxr-xr-x 2 root root 4096 May  9 14:05 /etc/ceph
  # however ceph-cluster runs as user 'ceph'
  # ceph-cluster         | uid=167(ceph) gid=167(ceph) groups=167(ceph)
  # So this helps correct it
  change-vol-ownership:
    # We can use any image we want as long as we can chown
    image: alpine
    # Priviliged enough to chown ;)
    user: "root"
    volumes:
      # The volume to chown
      - ceph-conf:/etc/ceph
    command: chown -R 167:167 /etc/ceph
  ceph-spdk:
```

Hello @epuertat 🖖,

Thank you for the impressive demo.

Trying to reproduce:
```shell
➜  ceph-nvmeof git:(build-container) ✗ make up                               
docker-compose up --detach --scale ceph-nvmeof=1 ceph-nvmeof
Creating network "ceph-nvmeof_default" with the default driver
Creating volume "ceph-nvmeof_ceph-conf" with default driver
Creating ceph-cluster ... done

ERROR: for ceph-nvmeof  Container "12dc650f465a" is unhealthy.
ERROR: Encountered errors while bringing up the project.
make: *** [up] Error 1
➜  ceph-nvmeof git:(build-container) ✗ make logs
docker-compose logs --follow --tail=40
Attaching to ceph-cluster
ceph-cluster            | grep: CMakeCache.txt: No such file or directory
ceph-cluster            | ceph-mgr dashboard not built - disabling.
ceph-cluster            | rm -f core* 
ceph-cluster            | hostname 12dc650f465a
ceph-cluster            | ip 172.25.0.2
ceph-cluster            | port 10000
ceph-cluster            | ./vstart.sh: line 670: /etc/ceph/ceph.conf: Permission denied
ceph-cluster exited with code 1
➜  ceph-nvmeof git:(build-container) ✗ 
```

While the following is a suggested workaround, there might be more elegant solutions available to resolve the issue. See also the second part bellow, triggering volume ownership change.

baum commented on docker-compose.yaml:
```diff
@@ -0,0 +1,70 @@
+    build:
+      context: .
+      dockerfile: Dockerfile.ceph
+    environment:
```
```suggestion
    depends_on:
      change-vol-ownership:
        condition: service_completed_successfully
    environment:
```

Second part of fix ceph conf directory ownership

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+COPY <<EOF /etc/yum.repos.d/ceph.repo 
+[Ceph]
+name=Ceph packages for \$basearch
+baseurl=http://download.ceph.com/rpm-$CEPH_VERSION/el\$releasever/\$basearch
```
```suggestion
baseurl=https://download.ceph.com/rpm-$CEPH_VERSION/el\$releasever/\$basearch
```

epuertat replied:
Good point! Thanks @ktdreyer. BTW this is literally taken from the repofile in https://download.ceph.com/rpm-17.2.6/el8/noarch/ceph-release-1-1.el8.noarch.rpm.

TBH, I'd love to replace get rid of this file inside the Dockerfile and go instead with a `RUN rpm -vih https://download.ceph.com/rpm-17.2.6/el8/noarch/ceph-release-1-1.el8.noarch.rpm`, but there's no EL9 release RPM and the EL8 doesn't include `el$releasever` variable, but it's pinned to `el8`.

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+baseurl=http://download.ceph.com/rpm-$CEPH_VERSION/el\$releasever/\$basearch
+enabled=1
+gpgcheck=1
+type=rpm-md
```
`type=rpm-md` is not needed on RHEL (I think this might be a SUSE thing)

epuertat replied:
Makes sense! As mentioned above, this also comes from our Ceph release RPM. Should we remove it there too?

epuertat replied:
BTW I also added the `priority=2` and `enabled=0` for the `ceph-source`s as suggested [here](https://docs.ceph.com/en/quincy/install/get-packages/#rhel).

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+# syntax = docker/dockerfile:1.4
+ARG OS_RELEASE=9
+FROM registry.access.redhat.com/ubi$OS_RELEASE/ubi:latest as prebase
```
It would be great to switch to ubi9-minimal (and microdnf) because that is what all the other product containers are using.

epuertat replied:
As mentioned, I can't fully replace this, because the `pkgdep.sh` script is external to this repo (`spdk` submodule) and relies on dnf (I should replace `dnf` with `microdnf` or alias it). That said, this is only used for the build stage, so my hope is that for the final stage we'll be able to use a lighter-weight base layer.

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+FROM registry.access.redhat.com/ubi$OS_RELEASE/ubi:latest as prebase
+ARG OS_RELEASE
+
+# Speed up dnf
```
"Speed up dnf" is not really clear to me - is it simply the `tsflags=nodocs`? 

If we switch to `ubi9-minimal` and `microdnf`, it's worth re-evaluating if each of these settings is still required.

epuertat replied:
> "Speed up dnf" is not really clear to me - is it simply the `tsflags=nodocs`?

TBH I didn't benchmark each tunable separately (I guess that `max_parallel_downloads` and `fastestmirror` could've also made a difference). I admit that it might be premature/inspecific optimization, but as I had to run this many, many times I wanted to reduce as much as possible time spent on `dnf` (the `pkgdep.sh` script invokes `dnf install` almost 30 times). The caching of `/var/cache/dnf` and `/var/lib/dnf` might also have helped ... I know it's ugly, so I'm open to remove it now it works.

> If we switch to `ubi9-minimal` and `microdnf`, it's worth re-evaluating if each of these settings is still required.

I'll try that. I was looking for a doc explaining differences between those and couldn't find any. I also tried installing from a centos-8/9 and the size increase was ~200 MB (from 300 to 500 MB), but it simplified setting the CentOS repos. I'll check what's the final size with those.

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+
+ARG CENTOS_URL="http://$CENTOS_MIRROR/$OS_RELEASE-stream/BaseOS/x86_64/os/Packages"
+ARG CENTOS_GPG_PACKAGE="$CENTOS_URL/centos-gpg-keys-$CENTOS_PACKAGE_VERSION.el$OS_RELEASE.noarch.rpm"
+ARG CENTOS_REPO_PACKAGE="$CENTOS_URL/centos-stream-repos-$CENTOS_PACKAGE_VERSION.el$OS_RELEASE.noarch.rpm"
```
This is complex to add `centos-stream-release` on to the ubi9 images. Some things I found that will lead to further maintenance:
- this uses plaintext http instead of https
- it uses pinned version numbers (`9.0-20` when `9.0-21` is out)
- `x86_64` is hard-coded into `CENTOS_URL`.  It's technically fine on non-x86 arches since the `centos-stream-repos` package is `noarch`, but that's a bit confusing on first reading

Maybe there's another way to approach this. What packages do you need from CentOS repos that are not available in the built-in ubi9 Yum repository?

I should probably try building this so I get a feel for it.

epuertat replied:
> This is complex to add `centos-stream-release` on to the ubi9 images. Some things I found that will lead to further maintenance:

Yeah, I don't like this too much either, but other approaches were even less maintainable (like directly installing remote RPMs)... I also tried using RockyLinux, since CentOS Stream is not the natural fit for UBI, but discarded that for different reasons.

> * this uses plaintext http instead of https

Unfortunately, CentOS mirrors [don't provide HTTPS support](https://serverfault.com/questions/1089416/why-are-centos-mirrors-http-and-not-https).

> * it uses pinned version numbers (`9.0-20` when `9.0-21` is out)

There's no way to solve this. Being a rolling release, sooner or later package versions + builds will change... (at least they keep ~3 older versions for these packages, so we won't need to update this immediately). Anyway, with a rolling release it's impossible to have an immutable/reproducible builds forever (unless we somehow snapshot the repository).

> * `x86_64` is hard-coded into `CENTOS_URL`.  It's technically fine on non-x86 arches since the `centos-stream-repos` package is `noarch`, but that's a bit confusing on first reading

Yeah, I was also surprised by this, but they don't provide a `noarch` dir :/

> Maybe there's another way to approach this. What packages do you need from CentOS repos that are not available in the built-in ubi9 Yum repository?
> 
> I should probably try building this so I get a feel for it.

Personally, for the build stage I'd use CentOS container as base image, since I'm not using the building stage for anything other than generating RPMs that are later copied on top of the final UBI base layer. I'll check what would be the missing packages in the final container.

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+ARG CENTOS_GPG_PACKAGE="$CENTOS_URL/centos-gpg-keys-$CENTOS_PACKAGE_VERSION.el$OS_RELEASE.noarch.rpm"
+ARG CENTOS_REPO_PACKAGE="$CENTOS_URL/centos-stream-repos-$CENTOS_PACKAGE_VERSION.el$OS_RELEASE.noarch.rpm"
+
+ARG CEPH_VERSION="17.2.5"
```
Can we avoid [hard-coding this version number](https://version.ninja/)? 17.2.6 is out

epuertat replied:
I initially left that blank and added a `RUN` assertion to fail if missing, but given that at some point we'll need to hardcode this string, I preferred to provide a working Dockerfile with no need to feed args/env. variables.

As a reference we're also 'hardcoding' the spdk release... I'd prefer to make it more explicit which packages and versions we are depending on. In an already closed PR to the `ceph/spdk` repo, I explored generating the SBOM, but that only works for attestations... Any ideas?

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+    --mount=type=cache,target=/root/.cache/pip \
+    dnf install -y \
+        rpm-build \
+        git \
```
The `git-core` RPM has fewer dependencies, and the main `git` RPM installs more scripts (that depend on Perl):
```suggestion
        git-core \
```

epuertat replied:
Good one! Thanks!

ktdreyer commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,107 @@
+EOF
+
+#------------------------------------------------------------------------------
+FROM prebase as base_release_8
```
You could reduce the complexity and drop CentOS 8, and only support CentOS 9

epuertat replied:
Makes sense. That said, I hoped to eventually convince the SPDK folks to add this Dockerfile to their project root, so we basically run a `docker build github.com/spdk/spdk` and we don't have to worry about maintaining that (they already have a Dockerfile, but it's for testing purposes).

baum commented on ceph-nvmeof.conf:
```diff
@@ -8,32 +8,29 @@
 name =
 group =
-addr = 127.0.0.1
+addr = ceph-nvmeof
```
@epuertat 🖖

I am not sure about the motivation for  `addr = ceph-nvmeof` and [`hostname = ceph-nvmeof`](https://github.com/epuertat/ceph-nvmeof/blob/b4d55a95cfbaf4fb9d01950c2baba4594ab51276/docker-compose.yaml#L45).
- What are we trying to achieve?
- What are disadvantages of using `addr = 0.0.0.0` and using automatically assigned hostnames based on their service names?

epuertat replied:
In this change I don't have yet your changes to the cli configuration, and I wanted to ensure that what I deliver here is fully functional. By using container hostnames we avoid the need of messing with IP addresses, and provide narrower exposure (`ceph-nvmeof` translates into 1 specific IP, which is valid both for server socket binding, and for client connection; but we don't bind to all interfaces, as with `0.0.0.0`). To be honest, this is not a big deal, since eventually Cephadm won't use any of this (this is only for development, testing and CI purposes).

epuertat commented:
Hey @ktdreyer, thank you very much for your thorough review! I already applied most of your suggestions, and I'm now checking with a CentOS 9 Stream for the build stage and ubi9 minimal for the runtime image.

BTW, unfortunately `microdnf` doesn't work yet (as per `ubi9/ubi-minimal:9.2-484` with [local installs](https://github.com/rpm-software-management/microdnf/issues/20). I'll check if that involves many deps, or I can handle it with rpm alone.

epuertat commented:
@ktdreyer I tried testing with `ubi-minimal` but it was impossible: `microdnf` doesn't support `localinstall`s which forces to use direct `rpm` instead, and then I had to manually install multiple deps (libnuma, libaio, ...), which btw are not available in the UBI repo, but in the CentOS one...

So then I finally preferred to stick to the (current) regular `ubi` base layer, and install the CentOS repo from the RPM packages. I could have also added the `centos.repo` and `centos-addons.repo` as files (as I did with the `ceph.repo`, as there`s no EL9 Ceph repo RPM), but those are big and I personally prefer to roll up the RPM package version on a yearly basis than hardcoding a 2 files from a foreign project (between the newest and the oldest `centos-stream-repos-9.0-21.el9.noarch.rpm` package there's more than a year).

Please, let me know your thoughts. Thanks!

epuertat commented:
BTW, @ktdreyer regarding:
* SPDK Container build time: I did some tests and got the following results (I decided to remove the `dnf.conf` tuning since 10-60 secs is																				 not that relevant, however `--mount=type=cache` brings ~50% seed-up):

|Case | Elapsed (secs) |
|-|-|								
|No tuning | 429 s|
| `--mount=type=cache` | 196 s (warm) - 310 s (cold) |
| `--mount=type=cache` + `dnf.conf` | 180 s (warm) - 235 s (cold) |

* Regarding the size: with the current approach (`ubi` + CentOS BaseOS) the resulting image was 342 MB (2 secs with warm cache). The `ubi-minimal` size is 300 MB (40 secs with warm cache).

I'll keep both (regular `ubi` and `minimal`, tunable with `CEPH_SPDK_BASE`), but until microdnf deals with local installs I don't think that the improvement from the minimal one is worth the extra effort.

baum commented on Dockerfile.spdk:
```diff
@@ -0,0 +1,85 @@
+#------------------------------------------------------------------------------
+FROM registry.access.redhat.com/ubi9/ubi AS ceph-spdk-ubi
+
+ARG CENTOS_BASE="https://mirror.stream.centos.org/9-stream/BaseOS/x86_64/os/Packages/"
```
One possible approach to avoid hardcoding `x86_64` could be to consider the following solution:
```suggestion
# Retrieve the machine architecture
RUN ARCH=$(uname -m)

ARG CENTOS_BASE="https://mirror.stream.centos.org/9-stream/BaseOS/$ARCH/os/Packages/"
```

baum commented:
@epuertat 🖖, one possible approach to avoid hardcoding `x86_64` in `Dockerfile.spdk` could be to consider [the following solution](https://github.com/ceph/ceph-nvmeof/commit/c5992658110e386cfa5262b0348683fb89f7e578).

epuertat commented:
> @epuertat :vulcan_salute:, one possible approach to avoid hardcoding `x86_64` in `Dockerfile.spdk` could be to consider [the following solution](https://github.com/ceph/ceph-nvmeof/commit/c5992658110e386cfa5262b0348683fb89f7e578).

Yeah, thanks for the idea! That's also used for installing the RPMs... The thing is that these packages are actually `noarch` (IMHO this is a questionable choice from the CentOS team, since platform-independent packages should be placed under `/noarch` dir), so there's no impact of hardcoding that (alternatively I thought about using some container predefined variables like `TARGETARCH`, but Docker uses the more appropriate `amd64` instead of `x86_64`).

baum reviewed (approved):
lgtm 🖖

gbregman commented:
@epuertat I'm trying to setup a cluster using the subnet we use for NVMe/RDMA and NVMe/TCP. I made these changes:

```
diff --git a/ceph-nvmeof.conf b/ceph-nvmeof.conf
index 84af075..68b0664 100644
--- a/ceph-nvmeof.conf
+++ b/ceph-nvmeof.conf
@@ -10,7 +10,7 @@
 [gateway]
 name =
 group =
-addr = 192.168.13.3
+addr = 172.168.24.3
 port = 5500
 enable_auth = False
 state_update_notify = True
diff --git a/docker-compose.yaml b/docker-compose.yaml
index 65fdce1..4a31e86 100644
--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -45,7 +45,7 @@ services:
       - ceph-conf:/etc/ceph
     networks:
       default:
-        ipv4_address: 192.168.13.2
+        ipv4_address: 172.168.24.2
   nvmeof-base:
     image: "quay.io/ceph/nvmeof:${VERSION?}"
     build:
@@ -78,7 +78,7 @@ services:
       - SYS_PTRACE # gdb
     networks:
       default:
-        ipv4_address: 192.168.13.3
+        ipv4_address: 172.168.24.3
     ports:
       - "4420:4420" # I/O controllers
       - "5500:5500" # Gateway
@@ -122,6 +122,6 @@ networks:
   default:
     ipam:
       config:
-        - subnet: 192.168.13.0/24
+        - subnet: 172.168.24.0/21
     labels:
       io.ceph.nvmeof: ""
```
But it doesn't work for me. I can't use the 172.168.24 subnet to run discover on another host. I can't even ping these addresses. Notice that before loading the container I can at least ping address 172.168.24.9, which is the host's address in this subnet but after loading the containers this ping stops working. Once I ran "make down" and stopped the containers, the ping started to work again.

epuertat commented:
> @epuertat I'm trying to setup a cluster using the subnet we use for NVMe/RDMA and NVMe/TCP. I made these changes:
> 
> ```
> diff --git a/ceph-nvmeof.conf b/ceph-nvmeof.conf
> index 84af075..68b0664 100644
> --- a/ceph-nvmeof.conf
> +++ b/ceph-nvmeof.conf
> @@ -10,7 +10,7 @@
>  [gateway]
>  name =
>  group =
> -addr = 192.168.13.3
> +addr = 172.168.24.3
>  port = 5500
>  enable_auth = False
>  state_update_notify = True
> diff --git a/docker-compose.yaml b/docker-compose.yaml
> index 65fdce1..4a31e86 100644
> --- a/docker-compose.yaml
> +++ b/docker-compose.yaml
> @@ -45,7 +45,7 @@ services:
>        - ceph-conf:/etc/ceph
>      networks:
>        default:
> -        ipv4_address: 192.168.13.2
> +        ipv4_address: 172.168.24.2
>    nvmeof-base:
>      image: "quay.io/ceph/nvmeof:${VERSION?}"
>      build:
> @@ -78,7 +78,7 @@ services:
>        - SYS_PTRACE # gdb
>      networks:
>        default:
> -        ipv4_address: 192.168.13.3
> +        ipv4_address: 172.168.24.3
>      ports:
>        - "4420:4420" # I/O controllers
>        - "5500:5500" # Gateway
> @@ -122,6 +122,6 @@ networks:
>    default:
>      ipam:
>        config:
> -        - subnet: 192.168.13.0/24
> +        - subnet: 172.168.24.0/21
>      labels:
>        io.ceph.nvmeof: ""
> ```
> 
> But it doesn't work for me. I can't use the 172.168.24 subnet to run discover on another host. I can't even ping these addresses. Notice that before loading the container I can at least ping address 172.168.24.9, which is the host's address in this subnet but after loading the containers this ping stops working. Once I ran "make down" and stopped the containers, the ping started to work again.

I guess that if you configure 2 different networks with the same range, that will probably break the routing. I'll be testing this on a couple of VMs.

epuertat committed:
build: add basic CI test

This adds a Github action that tests the basic commands and sets up
a minimal Ceph NVMe-oF gateway deployment.

Signed-off-by: Alexander Indenbaum <aindenba@redhat.com>
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>
Co-authored-by: Alexander Indenbaum <aindenba@redhat.com>

baum commented on Dockerfile:
```diff
@@ -0,0 +1,118 @@
+      io.ceph.git.repo="$NVMEOF_GIT_REPO" \
+      io.ceph.git.branch="$NVMEOF_GIT_BRANCH" \
+      io.ceph.git.commit="$NVMEOF_GIT_COMMIT"
+
```
```suggestion
RUN echo echo $NVMEOF_NAME $NVMEOF_VERSION $NVMEOF_GIT_COMMIT > /bin/ceph
RUN chmod 755 /bin/ceph
```

cephadm compatibility

Add `ceph` executable printing this component version, required by cephadm

cephadm `command_pull()`, during pulling the container images, calls to [command_inspect_image](https://github.com/ceph/ceph/blob/fd62a8dfcb345fabed3eea6b3a461f1ca848a239/src/cephadm/cephadm.py#L4995-L5011) which execs in container ["ceph --version"](https://github.com/ceph/ceph/blob/fd62a8dfcb345fabed3eea6b3a461f1ca848a239/src/cephadm/cephadm.py#L5007)
    
Example error:
```shell
RuntimeError: Failed command: /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --net=host --entrypoint ceph --init -e CONTAINER_IMAGE=quay.io/ceph/nvmeof:0.0.1 -e NODE_NAME=centos9 quay.io/ceph/nvmeof:0.0.1 --version: ERROR (catatonit:2): failed to exec pid1: No such file or directory
```

epuertat replied:
Thanks @baum!

Rather than creating a fake `ceph` binary, I'd add a `--version` flag support to the current entrypoint (`python3 -m control`), so that cephadm doesn't have to override the `entrypoint`:
```diff
- /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --net=host --entrypoint ceph --init -e CONTAINER_IMAGE=quay.io/ceph/nvmeof:0.0.1 -e NODE_NAME=centos9 quay.io/ceph/nvmeof:0.0.1 --version
+ /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --net=host --init -e CONTAINER_IMAGE=quay.io/ceph/nvmeof:0.0.1 -e NODE_NAME=centos9 quay.io/ceph/nvmeof:0.0.1 --version
```
@rkachach what do you think? Another approach?

baum replied:
> Thanks @baum!
> 
> Rather than creating a fake `ceph` binary, I'd add a `--version` flag support to the current entrypoint (`python3 -m control`), so that cephadm doesn't have to override the `entrypoint`:

@epuertat this is a common image pull path in cephadm, unrelated to nvmeof, the entry point is hardcoded as `ceph` see [here](https://github.com/ceph/ceph/blob/fd62a8dfcb345fabed3eea6b3a461f1ca848a239/src/cephadm/cephadm.py#L5007):

```python
    ver = CephContainer(ctx, ctx.image, 'ceph', ['--version']).run().strip()
```

rkachach replied:
@epuertat, as @baum pointed out this is a generic path in cephadm code and the boostrap process assumes that the image used is a `ceph based image` (so it must include the ceph binary).

epuertat replied:
@baum @rkachach as per our conversation I understand this wouldn't be required, right?

baum replied:
> @baum @rkachach as per our conversation I understand this wouldn't be required, right?

@epuertat 🖖,  The 'ceph' executable is necessary within the container image because the cephadm container management relies on its existence.

epuertat replied:
@baum, @rkachach told me otherwise, that this is only required for the bootstrap phase. If that were the case, we couldn't deploy Grafana, Prometheus or any other non-Ceph container...

baum replied:
@epuertat 🙂 running "ceph --version" is a step of the pull image procedure.  In a cephadm container management environment, if the 'ceph' executable is not present in the deployed image, that deployment would fail with a `RuntimeError: Failed command`.   Therefore 'ceph' executable is required for images targeted to be deployed by cephadm.

baum replied:
@rkachach 🖖, could you share your thoughts?

rkachach replied:
@epuertat @baum as I understood from @baum feedback the error was coming from the method:

```
    @staticmethod
    def get_version(ctx, container_id):
```

This method is meant to get the version of the container itself. In the case of iscsi for example it uses the version of the rpm package:
```
 '/usr/bin/python3 -c "import pkg_resources; print(pkg_resources.require('ceph_iscsi')[0].version)"
```

Not sure what would be the equivalent for nvmeof (nor if we will have an rpm package for the same or not).

As for the other issue, in principle ceph binary is not needed for daemons. It's called during bootstrap because this is normally run with the ceph image (which contains the binary obviously).

epuertat replied:
Good thing of the approach proposed by @baum (inspecting labels) is that it doesn't require to run the container, so it should be extremely efficient. If we wanted to keep consistency with other Ceph Python containers, we could do something like:
```
/usr/bin/python3 -c "print(__import__('control').__version__)""
```
But then we need to restore the `__version__ = x.y.z` symbol to the `__init__.py`. Given that we don't plan to release the NVMe-oF GW in other format than a container, I think that using `LABELs` and/or `ENV` variables is probably the best way.

epuertat replied:
BTW, our intention is that this `io.ceph.version` matches the container image tag (`nvmeof:<io.ceph.version>`), which will be hard-coded in the `cephadm` source code, right? Would still be necessary to dynamically discover that?

rkachach replied:
@epuertat awesome, it would be great if you provide me with a simple python code to get the version so I can use it as cephadm `def get_version()` method.

epuertat committed:
build: configure nofile

As the number of hosts connected to the NVMe-oF gateway increases, the
default nofile setting (max number of open file descriptors) of 1024
falls short, and needs to be increased.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: reduce hugepages

In Github CI hosts, requesting 4 GB of hugepages is not supported.
Reduced that to 1024 x 2 MB hugepages (2 GB).

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: fix hugepages and workflow

- Make .env bash-importable (add double quotes)
- Expose ports as .env settings (NVMEOF_GW_PORT, NVMEOF_IO_PORT,
  NVMEOF_DISC_PORT)
- Tune CI workflow to only run on push/pull-requests against 'devel'
  branch
- Add scheduled CI workflow run (nightly)
- Set hugepages in CI workflow to 256
- Fix issue with tput in non-TTY Github Action environment
- Use native GRPC_DNS_RESOLVER, as the ares one doesn't properly deal
  with Docker hostname resolution

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
cli: fix return code on error

- Return non-zero error code on client-server errors
- Add debug info to the CI job
- Fix & improve docs (SELinux config, typos)

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: force install thrift 0.14

Works-Around: https://tracker.ceph.com/issues/61882

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: debug Github Action with tmate

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: enable mlock in container

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: fix addr=0.0.0.0 issue

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: test mounting of nvmeof device

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: fix alias issue in Github Actions

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: disable nvme initiatior test

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
docs: update README

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: read .env in Makefile

This changes reads .env from Makefile which allows to simplify make
configuration with host networking.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: tune tcp options

Resolves: https://github.com/ceph/ceph-nvmeof/issues/160

Other changes:
- Move mk/demo.mk constants to .env
- Decrease Hugepages to 2 GB

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: add update-lockfile rule

When a Python package needs to be updated in pyproject.toml, the Python
package manager PDM will check that against the lockfile (pdm.lock) and
if any changes are found it will fail. This ensure builds are
reproducible and consistent.

In order to update, remove or add new packages, the lockfile needs to be
updated. A new Makefile target (update-lockfile) has been added. It runs
on a new container image (nvmeof-builder) and only updates the pdm.lock
file:

```bash
make update-lockfile
git add pdm.lock
```

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
doc: update dependencies

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
build: use ephemereal ports

Previously, docker-compose used fixed port mappings (host and container
ports matched, e.g: 4420:4420). However, that doesn't allow to deploy
multiple instances of the same container (--scale nvmeof=N).

Instead, ephemeral ports (randomly assigned by docker-compose) need to
be used. Fortunately, docker-compose provides with the 'port' command,
which maps container to host ports.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

oritwas reviewed (approved):
LGTM,
Thanks!
